src/librustc_typeck/check/closure.rs

```diff
@@ -155,6 +156,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             );
         }
 
+        // HACK(eddyb) this forces the types equated above into `substs` but
+        // it should rely on `ClosureSubsts` providing a constructor, instead.
+        let substs = self.resolve_vars_if_possible(&substs);
```

Comment:
the purpose of this code is to ensure that have a `fn` type, and not an inference variable that's been equated with a `fn` type?

Reply:
Yupp, pretty much.